curvefs/src/metaserver/register.cpp

```diff
@@ -0,0 +1,127 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: Thur May 9th 2019
+ * Author: lixiaocui
+ */
+
+#include <brpc/channel.h>
+#include <fcntl.h>
+#include <json2pb/json_to_pb.h>
+#include <json2pb/pb_to_json.h>
+
+#include <string>
+#include <vector>
+
+#include "curvefs/proto/topology.pb.h"
+#include "curvefs/src/metaserver/register.h"
+// #include "src/common/crc32.h"
+#include "src/common/string_util.h"
+#include "src/common/uri_parser.h"
+
+using ::curvefs::mds::topology::MetaServerRegistRequest;
+using ::curvefs::mds::topology::MetaServerRegistResponse;
+using ::curvefs::mds::topology::TopoStatusCode;
+
+namespace curvefs {
+namespace metaserver {
+Register::Register(const RegisterOptions &ops) {
+    this->ops_ = ops;
+
+    // 解析mds的多个地址
+    ::curve::common::SplitString(ops.mdsListenAddr, ",", &mdsEps_);
+    // 检验每个地址的合法性
+    for (auto addr : mdsEps_) {
+        butil::EndPoint endpt;
+        if (butil::str2endpoint(addr.c_str(), &endpt) < 0) {
+            LOG(FATAL) << "Invalid sub mds ip:port provided: " << addr;
+        }
+    }
+    inServiceIndex_ = 0;
+}
+
+int Register::RegisterToMDS(MetaServerMetadata *metadata) {
+    MetaServerRegistRequest req;
+    MetaServerRegistResponse resp;
+
+    req.set_hostip(ops_.metaserverInternalIp);
+    req.set_port(ops_.metaserverPort);
+    req.set_externalip(ops_.metaserverExternalIp);
+    req.set_externalport(ops_.metaserverPort);
+
+    LOG(INFO) << " Registering to MDS " << mdsEps_[inServiceIndex_]
+              << ". internal ip: " << ops_.metaserverInternalIp
+              << ", port: " << ops_.metaserverPort
+              << ", external ip: " << ops_.metaserverExternalIp;
+
+    int retries = ops_.registerRetries;
+    while (retries >= 0) {
+        brpc::Channel channel;
+        brpc::Controller cntl;
+
+        cntl.set_timeout_ms(ops_.registerTimeout);
+
+        if (channel.Init(mdsEps_[inServiceIndex_].c_str(), NULL) != 0) {
+            LOG(ERROR) << ops_.metaserverInternalIp << ":"
+                       << ops_.metaserverPort
+                       << " Fail to init channel to MDS "
+                       << mdsEps_[inServiceIndex_];
+            return -1;
+        }
+        curvefs::mds::topology::TopologyService_Stub stub(&channel);
+
+        stub.RegistMetaServer(&cntl, &req, &resp, nullptr);
+        if (!cntl.Failed() && resp.statuscode() == TopoStatusCode::TOPO_OK) {
+            break;
+        } else {
+            LOG(ERROR) << ops_.metaserverInternalIp << ":"
```

Comment:
LOG(INFO)

Reply:
> register fail, print error log

Register failed here will retry, LOG(INFO) here and LOG(ERROR) when out of retry times maybe better.